_posts/2020-07-05-raw-image-streaming.md

```diff
@@ -0,0 +1,125 @@
+We'll first dive into how Ironic deploys the images on the target hosts, and
+how raw image streaming improves this process. Afterwards, we will point out
+the changes to take this into use in Metal3.
+
+## Image deployments with Ironic
+
+In Metal3, the image deployment is performed by the Ironic Python Agent (IPA)
+image running on the target host. In order to deploy an image, Ironic will
+first boot the target node with an IPA image over iPXE. IPA will run in memory.
+
+Once IPA runs on the target node, Ironic will instruct it to download the
+target image. In Metal3, we use HTTP(S) for the download of the image. IPA will
+download the image and, depending on the format of the image, prepare it to
+write it on the disk. This means that the image is downloaded in memory and
+decompressed, two steps that can be both time and memory consuming.
+
+In order to improve this process, Ironic implemented a feature called raw image
+streaming.
+
+## What is raw image streaming ?
+
+The target image format when writing to disk is raw. That's why the images in
+formats like QCOW2 must be processed before being written to disk. However, if
+the image that is downloaded is already in raw format, then no processing is
+needed.
+
+Ironic leverages this, and instead of first downloading the image and then
+processing it before writing it to disk, it will directly write the
+downloading file to the disk. This feature is known as image streaming.
+Image streaming can only be performed with images in raw format.
+
+Since the downloaded image when streamed is directly written to disk, the
+memory size requirements changes. For any other format than raw, the target
+host needs to have sufficient memory to both run IPA (4GB) and
+download the image in memory. However with raw images, the only constraint
+on memory is to run IPA (so 4GB). For example, in order to deploy an Ubuntu
+image (around 700MB, QCOW2), the requirement is 8GB when in QCOW2 format, while
+it is only 4GB (as for any other image) when streamed as raw. This allows to
+deploy images that are bigger than the available memory on constrained nodes.
+
+However, this shifts the load on the network, since the raw images are usually
+much bigger than other formats. Using this feature in network constrained
+environment is not recommended.
+
+## Raw image streaming in Metal3
+
+In order to use raw image streaming in Metal3, a couple of steps are needed.
+The first one is to convert the image to raw and make it available in an
+HTTP server. This can be achieved by running :
+
+```bash
+qemu-img convert -O raw "${IMAGE_NAME}" "${IMAGE_RAW_NAME}"
+```
+
+Once converted the image format needs to be provided to Ironic through the
+BareMetalHost (BMH) image spec field. If not provided, Ironic will assume that
+the format is unspecified and download it in memory first.
+
+The following is an example of the BMH image spec field in Metal3-dev-env.
```

Comment:
it should be **m**etal3-dev-env in my opinion to maintain consistency

Reply:
or maybe Metal3 Dev Env ?